File: ui/api_client.py
"""
Thin HTTP client for the FastAPI backend (api/main.py). Keeps ui/app.py
free of raw `requests` calls and SSE-parsing details, and gives the UI a
small, easily-mocked surface for tests — mirrors how tools/stats_api.py
wraps football-data.org for the rest of the project.

Configure the backend's location with the API_BASE_URL environment
variable; defaults to http://localhost:8000 for local development.
"""

import json
import os
from typing import Callable, Optional

import requests

API_BASE_URL = os.getenv("API_BASE_URL", "http://localhost:8000")


class ApiError(Exception):
    """Raised when the backend can't be reached or returns an error."""


def get_competitions() -> list[dict]:
    try:
        r = requests.get(f"{API_BASE_URL}/competitions", timeout=10)
        r.raise_for_status()
        return r.json()
    except requests.RequestException as exc:
        raise ApiError(f"Could not reach the API at {API_BASE_URL}: {exc}") from exc


def get_teams(competition: Optional[str] = None) -> list[dict]:
    try:
        params = {"competition": competition} if competition else {}
        r = requests.get(f"{API_BASE_URL}/teams", params=params, timeout=30)
        r.raise_for_status()
        return r.json()
    except requests.RequestException as exc:
        raise ApiError(f"Could not reach the API at {API_BASE_URL}: {exc}") from exc


def get_crest(team_id: int) -> Optional[str]:
    """Never raises — a missing crest shouldn't block the page."""
    try:
        r = requests.get(f"{API_BASE_URL}/teams/{team_id}/crest", timeout=10)
        r.raise_for_status()
        return r.json().get("crest_url")
    except requests.RequestException:
        return None


def get_headline_stats(team_id: int, competition: str) -> Optional[dict]:
    """
    Current league-standing headline numbers for a team, or None if
    unavailable (e.g. a cup competition, or a request the backend can't
    fulfill) — never raises, since this is supplementary context for the
    hero section, not something that should block the page.
    """
    try:
        r = requests.get(
            f"{API_BASE_URL}/teams/{team_id}/headline", params={"competition": competition}, timeout=10,
        )
        r.raise_for_status()
        data = r.json()
        return data if data.get("available") else None
    except requests.RequestException:
        return None


def run_analysis(
    team_id: int,
    team_name: str,
    query: str,
    window_days: int,
    on_progress: Optional[Callable[[str], None]] = None,
) -> dict:
    """
    Calls the streaming /analyze/stream endpoint, invoking on_progress for
    each stage as it arrives, and returns the final result dict once the
    stream completes. Raises ApiError if the backend can't be reached, the
    pipeline itself fails, or the stream ends without ever sending a result.
    """
    payload = {"team_id": team_id, "team_name": team_name, "query": query, "window_days": window_days}
    try:
        response = requests.post(
            f"{API_BASE_URL}/analyze/stream", json=payload, stream=True, timeout=180,
        )
        response.raise_for_status()
    except requests.RequestException as exc:
        raise ApiError(f"Could not reach the API at {API_BASE_URL}: {exc}") from exc

    for raw_line in response.iter_lines(decode_unicode=True):
        if not raw_line or not raw_line.startswith("data: "):
            continue
        event = json.loads(raw_line[len("data: "):])

        if event["type"] == "progress" and on_progress:
            on_progress(event["stage"])
        elif event["type"] == "result":
            return event["result"]
        elif event["type"] == "error":
            raise ApiError(f"Analysis failed: {event['detail']}")

    raise ApiError("Stream ended without a result — the backend may have crashed mid-request.")

File: factpitch-frontend/src/services/apiClient.ts
/**
 * HTTP client for the FastAPI backend (api/main.py). Mirrors
 * ui/api_client.py's interface (same functions, same behavior) so the two
 * frontends stay easy to compare — the one real difference is
 * `runAnalysis`'s SSE consumption, which uses fetch + ReadableStream since
 * the browser has no equivalent of Python's `requests.iter_lines()`.
 */

import type { AnalyzeResult, Competition, HeadlineStats, StreamEvent, Team } from '../types/api';

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000';

export class ApiError extends Error {}

async function getJson<T>(path: string, params?: Record<string, string>): Promise<T> {
  const url = new URL(`${API_BASE_URL}${path}`);
  if (params) {
    for (const [key, value] of Object.entries(params)) {
      url.searchParams.set(key, value);
    }
  }

  let response: Response;
  try {
    response = await fetch(url);
  } catch (err) {
    throw new ApiError(`Could not reach the API at ${API_BASE_URL}: ${(err as Error).message}`);
  }

  if (!response.ok) {
    throw new ApiError(`Could not reach the API at ${API_BASE_URL}: ${response.status} ${response.statusText}`);
  }

  return response.json() as Promise<T>;
}

export async function getCompetitions(): Promise<Competition[]> {
  return getJson<Competition[]>('/competitions');
}

export async function getTeams(competition?: string | null): Promise<Team[]> {
  return getJson<Team[]>('/teams', competition ? { competition } : undefined);
}

/** Never throws — a missing crest shouldn't block the page. */
export async function getCrest(teamId: number): Promise<string | null> {
  try {
    const data = await getJson<{ crest_url: string | null }>(`/teams/${teamId}/crest`);
    return data.crest_url;
  } catch {
    return null;
  }
}

/** Never throws — headline stats are supplementary hero context, not
 * something that should block the page. Returns null both when the
 * backend says the data isn't available (e.g. a cup competition) and
 * when the request itself fails. */
export async function getHeadlineStats(teamId: number, competition: string): Promise<HeadlineStats | null> {
  try {
    const data = await getJson<HeadlineStats>(`/teams/${teamId}/headline`, { competition });
    return data.available ? data : null;
  } catch {
    return null;
  }
}

interface RunAnalysisParams {
  teamId: number;
  teamName: string;
  query: string;
  windowDays: number;
  onProgress?: (stage: string) => void;
  /** Optional AbortSignal so a caller (e.g. a React effect cleanup) can
   * cancel an in-flight analysis — useful if the user navigates away or
   * changes team mid-request. Not part of the Python client, since that's
   * a synchronous CLI/Streamlit context without an equivalent concern. */
  signal?: AbortSignal;
}

/**
 * Calls the streaming POST /analyze/stream endpoint, invoking onProgress
 * for each stage as it arrives, and resolves with the final result once
 * the stream completes. Throws ApiError if the backend can't be reached,
 * the pipeline itself fails, or the stream ends without ever sending a
 * result.
 */
export async function runAnalysis({
  teamId,
  teamName,
  query,
  windowDays,
  onProgress,
  signal,
}: RunAnalysisParams): Promise<AnalyzeResult> {
  let response: Response;
  try {
    response = await fetch(`${API_BASE_URL}/analyze/stream`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        team_id: teamId,
        team_name: teamName,
        query,
        window_days: windowDays,
      }),
      signal,
    });
  } catch (err) {
    throw new ApiError(`Could not reach the API at ${API_BASE_URL}: ${(err as Error).message}`);
  }

  if (!response.ok || !response.body) {
    throw new ApiError(`Could not reach the API at ${API_BASE_URL}: ${response.status} ${response.statusText}`);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = '';

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });

    // SSE events are separated by a blank line; a chunk boundary can land
    // mid-event, so only process complete lines and keep any trailing
    // partial line in the buffer for the next read.
    const lines = buffer.split('\n');
    buffer = lines.pop() ?? '';

    for (const rawLine of lines) {
      const line = rawLine.trim();
      if (!line || !line.startsWith('data: ')) continue;

      const event = JSON.parse(line.slice('data: '.length)) as StreamEvent;

      if (event.type === 'progress') {
        onProgress?.(event.stage);
      } else if (event.type === 'result') {
        return event.result;
      } else if (event.type === 'error') {
        throw new ApiError(`Analysis failed: ${event.detail}`);
      }
    }
  }

  throw new ApiError('Stream ended without a result — the backend may have crashed mid-request.');
}
